File: gnews-service/src/metrics.py
"""
Prometheus metrics for gnews-service
"""
from prometheus_client import Counter, Gauge, Histogram, CollectorRegistry

# Create a registry
registry = CollectorRegistry()

# Custom metrics
news_articles_fetched = Counter(
    'news_articles_fetched_total',
    'Total number of news articles fetched',
    ['status'],  # success/failure
    registry=registry
)

news_articles_published = Counter(
    'news_articles_published_total',
    'Total number of news articles published to RabbitMQ',
    registry=registry
)

news_fetch_duration = Histogram(
    'news_fetch_duration_seconds',
    'Time taken to fetch news from GNews API',
    buckets=[0.5, 1, 2, 5, 10, 30],
    registry=registry
)

news_api_errors = Counter(
    'news_api_errors_total',
    'Total number of errors from GNews API',
    ['error_type'],
    registry=registry
)

news_articles_in_db = Gauge(
    'news_articles_in_db',
    'Number of news articles currently in database',
    registry=registry
)

File: notifier-service/src/metrics.ts
/**
 * Prometheus metrics for notifier-service
 */
import client from 'prom-client';

// Create a Registry to register the metrics
export const register = new client.Registry();

// Add default metrics
client.collectDefaultMetrics({ register });

// Custom metrics
export const notificationsSent = new client.Counter({
  name: 'notifications_sent_total',
  help: 'Total number of notifications sent',
  labelNames: ['channel', 'status'], // channel: email/sms/whatsapp, status: success/failure
  registers: [register],
});

export const messagesConsumed = new client.Counter({
  name: 'rabbitmq_messages_consumed_total',
  help: 'Total number of messages consumed from RabbitMQ',
  labelNames: ['queue'],
  registers: [register],
});

export const sentimentAnalysisTotal = new client.Counter({
  name: 'sentiment_analysis_total',
  help: 'Total number of sentiment analysis performed',
  labelNames: ['sentiment'], // positive/negative/neutral
  registers: [register],
});

export const alertsTriggerTime = new client.Histogram({
  name: 'alerts_trigger_time_seconds',
  help: 'Time taken to trigger an alert',
  buckets: [0.1, 0.5, 1, 2, 5],
  registers: [register],
});
